src/AutoSchedule.cpp

```diff
@@ -0,0 +1,2312 @@
+                    if (i < num_pure_args && b.var == curr_f.args()[i])
+                        lower = Expr(b.min.as<IntImm>()->value);
+                }
+            }
+
+            if (!upper.as<IntImm>() && in_env) {
+                user_warning <<
+                "Bounds for the following expression could not be inferred" <<
+                "might result in suboptimal scheduling decisions:\n" << upper << '\n';
+                const Function &curr_f = env.at(f_reg.first);
+                for (auto &b: curr_f.schedule().estimates()) {
+                    uint32_t num_pure_args = curr_f.args().size();
+                    if (i < num_pure_args && b.var == curr_f.args()[i]) {
+                        const IntImm * bmin = b.min.as<IntImm>();
+                        const IntImm * bextent = b.extent.as<IntImm>();
+                        upper = Expr(bmin->value + bextent->value - 1);
+                    }
+                }
+            }
+
+            Interval concrete_bounds = Interval(lower, upper);
+            concrete_box.push_back(concrete_bounds);
+        }
+        concrete_regions[f_reg.first] = concrete_box;
+    }
+    return concrete_regions;
+}
+
+map<string, Box>
+DependenceAnalysis::redundant_regions(Function f, int stage_num, string var,
+                                      const DimBounds &bounds,
+                                      const set<string> &prods,
+                                      bool values_computed) {
+
+    map<string, Box> regions = regions_required(f, stage_num, bounds,
+                                                prods, values_computed);
+
+    DimBounds shifted_bounds;
+
+    for (auto &b : bounds) {
+        if (b.first == var) {
+            Expr len = b.second.max - b.second.min + 1;
+            Interval bound = Interval(b.second.min + len,
+                                      b.second.max + len);
+            shifted_bounds[b.first] = bound;
+        }
+        else {
+            shifted_bounds[b.first] = b.second;
+        }
+    }
+
+    map<string, Box> regions_shifted =
+            regions_required(f, stage_num, shifted_bounds, prods,
+                             values_computed);
+
+    map<string, Box> overalps;
+    for (auto &reg: regions) {
+        if (regions_shifted.find(reg.first) == regions.end()) {
+            // It will be interesting to log cases where this actually happens
+            // i.e., the shifted regions do not contain a function that was
+            // there in the original regions.
+            continue;
+        } else {
+            Box b = reg.second;
+            Box b_shifted = regions_shifted[reg.first];
+            // The boxes should be of the same size
+            internal_assert(b.size() == b_shifted.size());
+            // The box used makes things complicated ignoring it for now
+            Box b_intersect;
+            for (uint32_t i = 0 ; i < b.size(); i++)
```

Comment:
here and elsewhere: always wrap for loops in braces, even for single-line loops.